### DISCUSSION TITLE
Suggestions for better unlocked package creation

### DISCUSSION THREAD

**Discussion by madhav201098** [forcedotcom/cli#2513] [Q&A]
In our organization, we employ Unlocked packages for deploying changes to higher environments. We have observed that as the number of components continues to grow, the time required for creating a package version has been steadily increasing. Presently, it takes approximately 2.5 hours to create a single package version, and any errors encountered necessitate a restart. We are seeking recommendations on how to optimize the versioning of Unlocked packages to reduce the package creation time.

**Comment by mshanemc** [comment]
There's no great answers here, other than "the packaging and metadata API teams have some work in progress to improve the performance of package version creation"

Things you can do that would definitely help: 
- a larger number of smaller packages

Things that might help:
- improve apex test performance.  PVC is 4 phases (create org, deploy metadata, run tests, make package).  You can do the first 2 on a scratch org or just the 2nd one on a sandbox, and then run tests as a standalone command.  That would give you some benchmarks for how long the parts individual parts take.  And if it's tests that are slow, there are some approaches that make that go faster.

There's also a the skip-validation flag which helps if your package is not going to prod--you'd have to go back and make a "real" version of it once you know it's good.
`--skip-validation          -- Skip validation during package version creation; you can’t promote unvalidated package versions.`

**Comment by madhav201098** [comment]
Thanks @mshanemc for your response. 

1. is there any documentation or can you suggest how we can divide it into small packages? like which components can be divided etc
2. apart from the 4 steps mentioned, are there any other steps packaging commands run when it is initiated?
